### DISCUSSION TITLE
Why can I not read / write state (atom) values from functions directly?

### DISCUSSION THREAD

**Discussion by matthew-dean** [pmndrs/jotai#1422] [Ideas]
I've been evaluating different state libraries for React. I prefer Jotai over Valtio as it mimics `useState` instead of using side effects from Proxies. What's maddening, however, is that Jotai state cannot be used from utility functions, and I don't understand this. Why? Why not create a vanilla state library that has React effects (like Valtio does) instead of making React / components an implicit dependency? Why are these state values not readable / writeable? It's super-frustrating because there are lots of use cases for setting / reading state outside of React components, even within a React app. (Such as setting a global state for a set of utility functions.)

How hard would it be to add this functionality?

**Comment by dai-shi** [comment]
Thanks for opening a discussion. Jotai is based on React Context. A context is contextual and we only know where to access in a component. It's not accessible outside. 

Now, from your description, it seems you are not using the capability of context and it's fine. Many use cases don't require context and you are not alone. 

That's why Jotai provides the provider-less mode, which eases the non-context use cases. It also has unstable_createStore feature which allows you to access the store outside components. It's fairly stable but we would like to get more feedback before finalizing it. 

Hope it helps.

**Reply by matthew-dean** [reply]
In order to get more feedback, should you not publish it somewhere? I don't see `unstable_createStore` documented in Jotai documentation.

Also, on Github, it looks like it's been unstable for about a year? Or since 2021 anyway? Any idea when you might stabilize it (the API at least, so it wouldn't have the `unstable_` prefix? I think devs won't use it while it's called that).

**Reply by dai-shi** [reply]
> I don't see unstable_createStore documented in Jotai documentation.

Yeah, documentation would be required to get more feedback.

But, I mean feedback like this #1422 is fine and welcome. We are interested in use cases.

> Or since 2021 anyway?

https://github.com/pmndrs/jotai/releases/tag/v1.6.0
It's last Feb.

**Comment by dai-shi** [comment]
> We are interested in use cases.

So, why do you want to read/write atoms outside components/hooks?

**Reply by matthew-dean** [reply]
We're interested in a "simple" state solution for React and other utility functions where we _can_ use it in place for `useState` hooks, but don't have to.

Our main use case right now is a series of functions for converting dates / times, especially with timezones.

In the original implementation, a number of developers did something like this in a utility function:
```ts
export const convertTimeToDate = (time: string, date?: string | Date) => {
  const timeZone = Intl.DateTimeFormat().resolvedOptions().timeZone
  if (!date) {
    date = '2001-01-01'
  } else if (typeof date !== 'string') {
    date = format(date, 'YYYY-MM-dd')
  }
  return toDate(`${date}T${time}`, { timeZone })
}
```
In other words, we have a number of utility functions where we get the user's locale or time zone.

Eventually, we ran into a major problem. In our particular app, timezone / locale may not be a reliable constant, and certainly not reliable just by querying the browser. So we need to change these `resolvedOptions()` to _state_, and we need to get / set that state from DB values for a logged-in user.

Now, if we do it at the utility function level, it's super-trivial. We only need some basic refactoring to get the values from state (like Jotai or Valtio) instead of `resolvedOptions()`. So we can change it at the utility level, once. Like this:

```ts
const timeZone = store.get(timeZoneAtom)
export const convertTimeToDate = (time: string, date?: string | Date) => {
  if (!date) {
    date = '2001-01-01'
  } else if (typeof date !== 'string') {
    date = format(date, 'YYYY-MM-dd')
  }
  return toDate(`${date}T${time}`, { timeZone })
}
```

If, however, we were forced to do this via hooks, then we would need to change the API call from every single _use_ of every single one of these functions, in every single component, in every app. We would have to change the function signature for all functions, change how they were called, and increase the cognitive complexity of every component that calls these functions by adding an additional retrieval of `timeZone`. It's orders of magnitude more complex than just getting state from the store directly! The component doesn't actually _need_ the timezone in most use cases. (It might at points, which means we still need it to work like `useState`.) It just needs the utility function to know about the current timeZone of the current user, which we only need to set once per login.

**Reply by matthew-dean** [reply]
So like our case, it should be a common use case where a set of functions need state for things that are _mostly_ or _usually_ static, so passing that state "through" components is a completely worthless exercise since the component itself doesn't need the state at all.

Another example of this might be something like an auth token for an auth header for API requests. The component doesn't care at all about this header. It also has no place in calling API functions from components. Each component does not and _should_ not care about auth tokens! It would be 100% pure noise. But once that API call is called, it should include the current auth token, which it gets from state.

**Reply by dai-shi** [reply]
Thanks for the detailed use case. That's something I wanted to learn for various users.

Jotai (without unstable_createStore) is modeled after useState (+ useContext). It means state only exists in a component tree. This is important to isolate, for example, we can have multiple trees in memory. This is especially true with SSR. There exists multiple tress for clients/users. So, the fact that atom values are not accessible outside components is by design.

That said, there are many use cases that we only use one tree in memory. This is especially true in client/browser. 

Back to your use case, and auth token case, if they are single values in memory, we could use global variable.
```js
export let timeZone
export let authToken
```
So, when do we want them as atoms? It's when we need to render them in components. In such cases, we need atoms and something globally accessible. I would use `atomWithStore` from `jotai/zustand` for a simple case. `unstable_createStore` can do, but it's probably for more complex cases, like using react-three-fiber, or like we want to use 80% of atoms outside, apps like Excalidraw.

My apologies to make you get confused with `unstable_createStore`. I should have mentioned `atomWithStore` earlier.

**Reply by matthew-dean** [reply]
> So, when do we want them as atoms? It's when we need to render them in components.

Bear in mind that, in our use case, a user _could_ update their preferred timezone or locale. Meaning that anything on screen that is rendered using this same state should re-render.

I suppose, though, with our new flow, there isn't anything on screen so it could just be a simple object with a getter / setter. What I didn't articulate is I was looking for something that's a "general shared state" solution, so ideally we would use the same state library / pattern for utility state as we'd use for shared component state.

**Reply by dai-shi** [reply]
> Meaning that anything on screen that is rendered using this same state should re-render.

Gotcha.

> so ideally we would use the same state library / pattern for utility state as we'd use for shared component state.

I understand where it's coming from. And, yeah, global state (like zustand/valtio) vs. component state (like jotai or `useState`) is a tough one. I think people would just use global state if they need to read state outside components. with `useState`, people may use `useEffect` but my recommendation is to update component state and global variable at the same time.

```js
let globalToken = '';
const useToken = () => {
  const [token, setToken] = useState(globalToken);
  return [
    token,
    (nextToken) => { setToken(nextToken); globalToken = nextToken; }
  ];
};
```

(I can continue with some jotai patterns, but I think our discussion now is something general.)

**Reply by matthew-dean** [reply]
@dai-shi Here's another use case -- for LaunchDarkly, again we want feature flags available to normal functions but also be reactive / observable to components. LaunchDarkly supplies a React SDK, but the React SDK _omits_ a JavaScript sdk. So I wrote this code today to be able to serve feature flags in functions and components.

```tsx
import * as LDClient from 'launchdarkly-js-client-sdk'
import { atomWithStorage } from 'jotai/utils'
import { useState } from 'react'

import store from './store'

/**
 * Will persist feature flags to localStorage / sessionStorage,
 * depending on what is availabile
 */
const featureFlagsAtom = atomWithStorage('featureFlags', {})

/**
 * We do this instead of `useAtom` so that we make sure we're
 * using the same store for setting / getting values.
 */
const useFeatureFlags = () => {
  const [flags, setFlags] = useState(store.get(featureFlagsAtom))
  store.sub(featureFlagsAtom, () => {
    setFlags(store.get(featureFlagsAtom))
  })
  return flags
}

const setupFeatureFlags = (clientId: string) => {
  const client = LDClient.initialize(clientId, {}, {
    sendEventsOnlyForVariation: true
  })
  client.on('initialized', () => {
    store.set(featureFlagsAtom, client.allFlags())
  })
}
```

**Reply by gregor-mueller** [reply]
I also got a use case where I need the atom to be reactive but also available in a non-react function.

What would be your proposal for making this work?

I was expecting (using v2 API) `store.get(myAtom)` to return the latest value but it is always returning the initial value until I explicitly update it using `store.set(myAtom, newValue)`.  
Is this working as intended? It feels like a bug, to be honest.  
If it is intended though, I guess calling `store.set` in a `useEffect` would be the way to go, wouldn't it?

Here is a reproduction link:
https://codesandbox.io/s/jotai-store-get-xqidvn?file=/src/App.js


If you want, I already have a issue prepared to be created (the checklist from the template actually made me look for the discussions first 😄 ).

**Reply by dai-shi** [reply]
> (the checklist from the template actually made me look for the discussions first 😄 ).

Thanks!

`import {} from 'jotai'` is v1 API. You have to use `import {} from 'jotai/vanilla'`, `/react`, or so for v2 API.

(It will be fine after `jotai@2.0.0` is released.)

**Reply by gregor-mueller** [reply]
I see, it's working then.

Maybe you want to update the docs then to prevent further confusion, until v2 is released.

The docs mention importing from `jotai` directly:  
![image](https://user-images.githubusercontent.com/28727932/211272748-e187522b-6278-4e8f-9aed-0b0798177aa4.png)

**Reply by dai-shi** [reply]
I should've said "they will be the defaults"..

**Reply by gregor-mueller** [reply]
> I see, it's working then.

...in the sandbox example.
In our production code, we also use

```ts
import { atomFamily } from 'jotai/utils'
```

which I replaces successfully with

```ts
import { atomFamily } from 'jotai/vanilla/utils'
```

An issue that I was not able to resolve myself is using

```ts
import { withImmer } from 'jotai/immer'
```

I replaced it with

```ts
import { withImmer } from 'jotai-immer'
```
I installed this version: https://ci.codesandbox.io/status/jotai-labs/jotai-immer/pr/1/builds/323416

It does have a typing issue with `atomFamily` in combination with `useAtomCallback` and `withImmer`.

(note: We prefix all atoms with `$` instead of suffixing with `Atom`.)

```ts
type Item = {
  id: string
  parentId: string
}

const $items = withImmer(atom<Item[]>([]))

const $item = atomFamily((id: string) =>
  withImmer(
    atom(
      (get) => get($items).find((candidate) => candidate.id === id),
      (get, set, value: Item | undefined) => {
        set($items, (items) => {
          if (value) {
            items[items.findIndex((item) => item.id === id)] = value
          }
        })
      },
    ),
  ),
)

const $openItemId = atom<string | undefined>(undefined)

const foo = useAtomCallback((get, set) => {
  const openItemId = get($openItemId) // This is where it throws
})
/*
TS2769: No overload matches this call.
  Overload 1 of 6, '(atom: Atom<string | Promise<string | undefined> | undefined>): string | undefined', gave the following error.
    Argument of type 'PrimitiveAtom<string | undefined> & WithInitialValue<string | undefined>' is not assignable to parameter of type 'Atom<string | Promise<string | undefined> | undefined>'.
      Types of property 'read' are incompatible.
        Type 'Read<string | undefined, SetAtom<[SetStateAction<string | undefined>], void>>' is not assignable to type 'Read<string | Promise<string | undefined> | undefined>'.
  Overload 2 of 6, '(atom: Atom<Promise<unknown>>): unknown', gave the following error.
    Argument of type 'PrimitiveAtom<string | undefined> & WithInitialValue<string | undefined>' is not assignable to parameter of type 'Atom<Promise<unknown>>'.
      Types of property 'read' are incompatible.
        Type 'Read<string | undefined, SetAtom<[SetStateAction<string | undefined>], void>>' is not assignable to type 'Read<Promise<unknown>>'.
  Overload 3 of 6, '(atom: Atom<string | undefined>): string | undefined', gave the following error.
    Argument of type 'PrimitiveAtom<string | undefined> & WithInitialValue<string | undefined>' is not assignable to parameter of type 'Atom<string | undefined>'.
      Types of property 'read' are incompatible.
        Type 'Read<string | undefined, SetAtom<[SetStateAction<string | undefined>], void>>' is not assignable to type 'Read<string | undefined>'.
*/
```

So it is adding the `Promise` type which the `Getter` in `useAtomCallback` does not seem to like

**Reply by dai-shi** [reply]
can't you reproduce the error in codesandbox?

**Reply by dai-shi** [reply]
btw, thanks for trying v2 API. we need feedback before finalizing it.

**Reply by gregor-mueller** [reply]
> can't you reproduce the error in codesandbox?

Yes! After tinkering around, I found it to be almost working.

2 issues left.


### Issue 1
The first issue that I see left right now is the typing of the setter function of an atom with initialiser that can also be undefined (will be `never`):

```ts
const selectedNumber = atom<number | undefined>(undefined)

const SomeThing(id: number) {
  const setSelected = useSetAtom(selectedNumber);

  return <button onClick={() => setSelected(id)}>Select {id}</button>
}
```

`setSelected` will be typed as `never` but actually work as a setter.

Link to affected line sandbox:  
https://codesandbox.io/s/jotai-store-get-with-atomcallback-h8c9rf?file=/src/Counts.tsx:704-793


### Issue 2

It is not possible to use an `atomFamily` with `store.get(myFamilyAtom)`

Link to affected line sandbox:  
https://codesandbox.io/s/jotai-store-get-with-atomcallback-h8c9rf?file=/src/notifications.ts:308-377

**Reply by dai-shi** [reply]
Re: Issue 1

That happens when `strictNullChecks: false` in tsconfig.json. In general, `strict: true` is recommended, but a workaround would be this:
```ts
export const $selectedCountId = atom(undefined as number | undefined);
```

Re: Issue 2

It should be this:
```ts
  const count = jotaiStore.get($count(selectedCountId));
```

You might misunderstand atomFamily. It's just a Map:
```js
// simplified code
const atomFamily = (initializeAtom) => {
  const cache = new Map();
  return (param) => {
    if (!cache.has(param)) {
      cache.set(param, initializeAtom(param));
    }
    return cache.get(param);
  };
};
```
ref: https://github.com/pmndrs/jotai/discussions/1323#discussioncomment-3303035

**Reply by gregor-mueller** [reply]
> Re: Issue 1
> 
> That happens when `strictNullChecks: false` in tsconfig.json. In general, `strict: true` is recommended, but a workaround would be this:
> 
> ```ts
> export const $selectedCountId = atom(undefined as number | undefined);
> ```

I see, I thought, it was using `strictNullChecks: true` by default, as that's the config when you "create" a `tsconfig.json` in codesandbox.

👍 

> Re: Issue 2
> 
> It should be this:
> 
> ```ts
>   const count = jotaiStore.get($count(selectedCountId));
> ```
> 
> You might misunderstand atomFamily. It's just a Map:
> 
> ref: [#1323 (reply in thread)](https://github.com/pmndrs/jotai/discussions/1323#discussioncomment-3303035)

Gotcha! 👍 


Thanks for responding in a quick and helpful manner!

**Comment by MOZGIII** [comment]
An example from the real world of how this can be useful:

I have a suite of helper functions (technically it is a service model, but we can say helper functions) that represents an action that spans over many atoms.

My app is a time tracker, and a good example is what happens when the user presses a stop button on the currently tracked entry.
The logic is something like this:
1. Call the `trackerActions.stopAndRecord` helper and pass the `currentTime` to it, which will internally do:
	1. Get the `trackingState` and `taskName` atom values
	2. Call the `trackerActions.stop` helper, which will internally do:
		1. Reset the `trackingState` and `taskName` atoms
	3. Construct a new `record` object with the values of `taskName` and `trackingState` atoms that we have captured captured before and the current time
	4. Call the `recordsActions.upsert` helper passing the `record`, which will internally do:
		1. Update the value of the `records` atom inserting or updating the record entry

The problem here is that the code that invokes the actions does not have anything to do with records (i.e. it is a timer stop button), but it would require importing the records action to proceed. In addition to that, it would be necessary to pass the `setRecords` to both the `recordsActions.upsert` (reasonable) and `trackerActions.stopAndRecord` (eh... ok?). My concern is what would happen if the helper is more complicated - passing one contextual-information carrting parameters is fine, two - a bit too much, but for each of the accessed atom would be painful for the ergonomics IMO.

---

I suggest `jotai` expose the underlying state store more explicitly when needed. I have two suggestions:

1. `useAtomsStore`
   	```ts
	// Declare a global atom
	const myAtom = atom(0);

    // Declare a global helper
	const myhelper = (atomsStore: ReturnType<typeof createStore>) => {
        myAtom.write(atomsStore.get, atomsStore.set, 1); // set the value to `1`
    }

    // In react
	const atomsStore = useAtomsStore(); // captures the context
	myHelper(atomsStore); // invoke the helper passing access to the store
   	```

2. `useGetAtomStore`
   	```ts
	// Declare a global atom
	const myAtom = atom(0);

    // Declare a global helper
	const myhelper = (getAtomStore: (atom: Atom<any>) => ReturnType<typeof createStore>) => {
        const storeForAParticularAtom = getAtomStore(myAtom);
        // getAtomStore(myOtherAtom) might return a different value here,
		// and this opportunity to return different stores per atom is the key difference to the first suggestion 
        myAtom.write(storeForAParticularAtom.get, storeForAParticularAtom.set, 1); // set the value to `1`
    }

    // In react
	const getAtomStore = useGetAtomStore(); // captures the context
	myHelper(getAtomStore); // invoke the helper passing access to the store
   	```

In both cases we can capture the whatever store it currently utilized by react via a hook, so we don't loose the notion of the context-based state.
For global access, i.e. when the action truly has to originate from outside of react, it would be possible to create a global `state` (via `createState`) first, and then attach it to the tree via `Provider` so react uses it, and then pass it to the helpers directly when the callside does not have react contextual tree information.
Alternatively, the helper component can be inserted into the tree to expose the state information at a particular point via a global variable (`let atomsStore;` globally and `atomsStore = useAtomsStore()`) - but this is actually a mess because of concurrent rendering and remounts...

**Reply by dai-shi** [reply]
In case you don't follow, Jotai v2 API will expose store API. #1514

**Reply by MOZGIII** [reply]
I can see that, great!
